fix(matrices): Don't use DomainMatrix.rref for RR in linsolve

oscarbenjamin opened the pull request:
The DomainMatrix.rref method as implemented by the sdm_irref function
has not been well designed for handling matrices of floats as it does
not have a pivoting strategy that can mitigate errors arising from
erroneously choosing a zero pivot (that is nonzero only because of
roundoff). This commit makes linsolve use the `ddm_irref()` for the case of
pure RR or CC matrices. The behaviour of sdm_rref for floats should be
improved. Until then it is better not to use `sdm_rref` in these
cases.

<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->

Fixes #21520

#### Brief description of what is fixed or changed


#### Other comments


#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* solvers
    * Some bugs in the new implementation of linsolve are fixed by the dense RREF implementation instead of the sparse implemention. These bugs previously led to incorrect results being returned linsolve.
<!-- END RELEASE NOTES -->

oscarbenjamin committed:
fix(matrices): Don't use DomainMatrix.rref for RR in linsolve

The DomainMatrix.rref method as implemented by the sdm_irref function
has not been well designed for handling matrices of floats as it does
not have a pivoting strategy that can mitigate errors arising from
erroneously choosing a zero pivot (that is nonzero only because of
roundoff). This commit makes linsolve use Matrix.rref() for the case of
pure float matrices. The behaviour of sdm_rref for floats should be
improved. Until then it is better not to use DomainMatrix in these
cases.

jksuom commented on sympy/polys/matrices/linsolve.py:
```diff
@@ -73,6 +74,13 @@ def _linsolve(eqs, syms):
+    # sdm_irref has issues with float matrices. This uses the Matrix.rref()
+    # method. When DomainMatrix.rref() can handle float matrices reasonably
+    # this should be removed...
+    if K.is_RealField:
```
Complex field should treated similarly.

oscarbenjamin replied:
Yes, it should although it isn't actually used here since it always goes to EX:
```python
In [2]: construct_domain([1.0+1.0*I])
Out[2]: (EX, [EX(1.0 + 1.0*I)])

In [3]: Poly(1.0+1.0*I, x)
Out[3]: Poly(1.0 + 1.0*I, x, domain='EX')
```
I think the only way to get CC is by unifying `ZZ_I/QQ_I` with `RR`:
```python
In [4]: ZZ_I.unify(RR)
Out[4]: CC
```

oscarbenjamin replied:
I've changed it so that `construct_domain` can identify `CC`. I would have expected to see some test failures from that but there don't seem to be any which means that there is not good test coverage for Poly/linsolve etc over CC.

It should be possible to test `if K.is_Exact` here but that apparently fails for a polynomial ring:
```python
In [1]: RR.frac_field(x).is_Exact
Out[1]: False

In [2]: RR[x].is_Exact
Out[2]: True
```

oscarbenjamin committed:
fix(polys): make construct_domain use CC for 1.0+1.0*I

Previously construct_domain would give EX for complex floats like
1.0+1.0*I so that the CC domain is not really used in cases like
Poly(1.0+1.0*I, x)

oscarbenjamin commented:
I'm going to try and implement partial pivoting for RR so I will follow up on this.

Also there should be `CC` tests for things like `Poly(1.0+I, x)`.

Presumably it's also possible for these issues to arise when dealing with a domain like `RR(x)` but I'm not sure how pivoting would work there. For RR you find the element with largest absolute value so for `RR(x)` would you use the leading coefficient of the numerator? Maybe there are issues with having e.g. a range of magnitudes in the same polynomial like `1e-10*x**2 + 1e10*x + 1e-20`...

Otherwise we should make sure to merge this as a workaround before sympy 1.9.

oscarbenjamin committed:
test(polys): make linsolve tests more robust

oscarbenjamin committed:
test(polys): test constructing the CC domain

oscarbenjamin committed:
fix(polys): use DDM instead of Matrix for linsolve with RR/CC

Adds partial pivoting to DDM.rref and uses DDM.rref in linsolve for the
RR and CC domains. This is to avoid the problems with sdm_rref which
does not have adequate pivoting to handle numeric instability.

oscarbenjamin committed:
fix(polys): remove unneeded import

oscarbenjamin committed:
test(polys): add comment about difficult system in linsolve tests

oscarbenjamin commented:
I think this is good for review/merge now.

There is still more work to be done on this but the basic bug is fixed and there are now some floating point tests.

jksuom commented:
Looks good to me.

oscarbenjamin commented:
Thanks.

I've been thinking about `RR(x)` and `CC(x)` and I don't think that they are practically usable domains for `linsolve`. I thought that maybe `solve` had a better approach to this but I realise now that it only works because of `rational=True` which uses `nsimplify`:
```python
In [1]:     eqs = [
   ...:         x*(0.2*z + 0.9) + y*(0.5*z + 0.8) + 0.6,
   ...:         0.1*x*z + y*(0.1*z + 0.6) + 0.9,
   ...:     ]

In [2]: eqsz0 = [eq.subs(z, 0) for eq in eqs]

In [3]: linsolve(eqs, [x, y])
Out[3]: 
⎧⎛          2                                                                   ⎞⎫
⎪⎜- 3.9e-5⋅z  - 3.6e-5⋅z - 8.67361737988404e-20        3.6e-5⋅z + 0.000243      ⎟⎪
⎨⎜─────────────────────────────────────────────, ───────────────────────────────⎟⎬
⎪⎜               3           2                           2                      ⎟⎪
⎩⎝       3.0e-6⋅z  - 1.3e-5⋅z  - 5.4e-5⋅z        9.0e-6⋅z  - 3.9e-5⋅z - 0.000162⎠⎭

In [4]: linsolve(eqs, [x, y]).subs(z, 0)
Out[4]: {(zoo, -1.5)}

In [5]: eqsz0
Out[5]: [0.9⋅x + 0.8⋅y + 0.6, 0.6⋅y + 0.9]

In [6]: linsolve(eqsz0, [x, y])
Out[6]: {(0.666666666666667, -1.5)}

In [7]: eqsQQ = [nsimplify(eq) for eq in eqs]

In [8]: linsolve(eqsQQ, [x, y])
Out[8]: 
⎧⎛   -39⋅z - 36        12⋅z + 81    ⎞⎫
⎪⎜────────────────, ────────────────⎟⎪
⎨⎜   2                 2            ⎟⎬
⎪⎝3⋅z  - 13⋅z - 54  3⋅z  - 13⋅z - 54⎠⎪
⎩                                    ⎭

In [9]: linsolve(eqsQQ, [x, y]).subs(z, 0)
Out[9]: {(2/3, -3/2)}

In [10]: solve(eqs, [x, y])
Out[10]: 
⎧       -39.0⋅z - 36.0             12.0⋅z + 81.0     ⎫
⎪x: ──────────────────────, y: ──────────────────────⎪
⎨        2                          2                ⎬
⎪   3.0⋅z  - 13.0⋅z - 54.0     3.0⋅z  - 13.0⋅z - 54.0⎪
⎩                                                    ⎭

In [11]: solve(eqs, [x, y], rational=False)
Out[11]: 
⎧          ⎛           4          3            2                      ⎞               ⎛       2                 ⎞   ⎫
⎪      1.0⋅⎝- 0.00312⋅z  - 0.045⋅z  - 0.22842⋅z  - 0.45927⋅z - 0.26244⎠           1.0⋅⎝0.024⋅z  + 0.27⋅z + 0.729⎠   ⎪
⎨x: ──────────────────────────────────────────────────────────────────────, y: ─────────────────────────────────────⎬
⎪            5           4            3            2                                  3          2                  ⎪
⎩   0.00024⋅z  + 0.0022⋅z  - 0.00378⋅z  - 0.09963⋅z  - 0.35721⋅z - 0.39366     0.006⋅z  + 0.001⋅z  - 0.225⋅z - 0.486⎭

In [12]: from sympy.polys.matrices import DomainMatrix

In [13]: A, b = linear_eq_to_matrix(eqs, [x, y])

In [14]: Aaug = Matrix.hstack(A, b)

In [15]: Aaug
Out[15]: 
⎡0.2⋅z + 0.9  0.5⋅z + 0.8  -0.6⎤
⎢                              ⎥
⎣   0.1⋅z     0.1⋅z + 0.6  -0.9⎦

In [16]: Aaug.rref()
Out[16]: 
⎛⎡                    2                      ⎤        ⎞
⎜⎢           - 0.078⋅z  - 0.423⋅z - 0.324    ⎥        ⎟
⎜⎢1  0  ─────────────────────────────────────⎥        ⎟
⎜⎢             3          2                  ⎥        ⎟
⎜⎢      0.006⋅z  + 0.001⋅z  - 0.225⋅z - 0.486⎥        ⎟
⎜⎢                                           ⎥, (0, 1)⎟
⎜⎢                  0.12⋅z + 0.81            ⎥        ⎟
⎜⎢0  1         ───────────────────────       ⎥        ⎟
⎜⎢                   2                       ⎥        ⎟
⎝⎣             0.03⋅z  - 0.13⋅z - 0.54       ⎦        ⎠

In [17]: Aaugdm = DomainMatrix.from_Matrix(Aaug)

In [18]: A.inv()
Out[18]: 
⎡              2                                     2                       ⎤
⎢      - 0.02⋅z  - 0.21⋅z - 0.54                0.1⋅z  + 0.61⋅z + 0.72       ⎥
⎢─────────────────────────────────────  ─────────────────────────────────────⎥
⎢       3          2                           3          2                  ⎥
⎢0.006⋅z  + 0.001⋅z  - 0.225⋅z - 0.486  0.006⋅z  + 0.001⋅z  - 0.225⋅z - 0.486⎥
⎢                                                                            ⎥
⎢                0.1⋅z                               -0.2⋅z - 0.9            ⎥
⎢       ───────────────────────                ───────────────────────       ⎥
⎢             2                                      2                       ⎥
⎣       0.03⋅z  - 0.13⋅z - 0.54                0.03⋅z  - 0.13⋅z - 0.54       ⎦

In [21]: Aaugdm[:,:2].to_field().inv().to_Matrix()
Out[21]: 
⎡                 4           3           2                                           2                       ⎤
⎢       - 0.0008⋅z  - 0.0156⋅z  - 0.1134⋅z  - 0.3645⋅z - 0.4374                  0.1⋅z  + 0.61⋅z + 0.72       ⎥
⎢──────────────────────────────────────────────────────────────────────  ─────────────────────────────────────⎥
⎢         5           4            3            2                               3          2                  ⎥
⎢0.00024⋅z  + 0.0022⋅z  - 0.00378⋅z  - 0.09963⋅z  - 0.35721⋅z - 0.39366  0.006⋅z  + 0.001⋅z  - 0.225⋅z - 0.486⎥
⎢                                                                                                             ⎥
⎢                                 2                                                                           ⎥
⎢                           0.02⋅z  + 0.09⋅z                                          -0.2⋅z - 0.9            ⎥
⎢                ─────────────────────────────────────                          ───────────────────────       ⎥
⎢                       3          2                                                  2                       ⎥
⎣                0.006⋅z  + 0.001⋅z  - 0.225⋅z - 0.486                          0.03⋅z  - 0.13⋅z - 0.54       ⎦

In [22]: Aaugdm.to_field().rref()[0].to_Matrix()
Out[22]: 
⎡                     2                                   ⎤
⎢          - 0.00039⋅z  - 0.00036⋅z - 8.67361737988404e-19⎥
⎢1.0   0   ───────────────────────────────────────────────⎥
⎢                         3            2                  ⎥
⎢                 3.0e-5⋅z  - 0.00013⋅z  - 0.00054⋅z      ⎥
⎢                                                         ⎥
⎢                         0.012⋅z + 0.081                 ⎥
⎢ 0   1.0            ──────────────────────────           ⎥
⎢                           2                             ⎥
⎣                    0.003⋅z  - 0.013⋅z - 0.054           ⎦

In [23]: Aaugdm.to_dense().to_field().rref()[0].to_Matrix()
Out[23]: 
⎡                           4          3            2                            ⎤
⎢                - 0.00312⋅z  - 0.045⋅z  - 0.22842⋅z  - 0.45927⋅z - 0.26244      ⎥
⎢1.0   0   ──────────────────────────────────────────────────────────────────────⎥
⎢                   5           4            3            2                      ⎥
⎢          0.00024⋅z  + 0.0022⋅z  - 0.00378⋅z  - 0.09963⋅z  - 0.35721⋅z - 0.39366⎥
⎢                                                                                ⎥
⎢                                       2                                        ⎥
⎢                                0.024⋅z  + 0.27⋅z + 0.729                       ⎥
⎢ 0   1.0                  ─────────────────────────────────────                 ⎥
⎢                                 3          2                                   ⎥
⎣                          0.006⋅z  + 0.001⋅z  - 0.225⋅z - 0.486                 ⎦
```
If we compute the adjugate and determinant then we should be able to get something better:
```python
In [25]: A.adjugate() / A.det()
Out[25]: 
⎡       0.1⋅z + 0.6                -0.5⋅z - 0.8      ⎤
⎢─────────────────────────  ─────────────────────────⎥
⎢        2                          2                ⎥
⎢- 0.03⋅z  + 0.13⋅z + 0.54  - 0.03⋅z  + 0.13⋅z + 0.54⎥
⎢                                                    ⎥
⎢         -0.1⋅z                   0.2⋅z + 0.9       ⎥
⎢─────────────────────────  ─────────────────────────⎥
⎢        2                          2                ⎥
⎣- 0.03⋅z  + 0.13⋅z + 0.54  - 0.03⋅z  + 0.13⋅z + 0.54⎦

In [26]: A.inv()
Out[26]: 
⎡              2                                     2                       ⎤
⎢      - 0.02⋅z  - 0.21⋅z - 0.54                0.1⋅z  + 0.61⋅z + 0.72       ⎥
⎢─────────────────────────────────────  ─────────────────────────────────────⎥
⎢       3          2                           3          2                  ⎥
⎢0.006⋅z  + 0.001⋅z  - 0.225⋅z - 0.486  0.006⋅z  + 0.001⋅z  - 0.225⋅z - 0.486⎥
⎢                                                                            ⎥
⎢                0.1⋅z                               -0.2⋅z - 0.9            ⎥
⎢       ───────────────────────                ───────────────────────       ⎥
⎢             2                                      2                       ⎥
⎣       0.03⋅z  - 0.13⋅z - 0.54                0.03⋅z  - 0.13⋅z - 0.54       ⎦
```
The method for computing adjugate here in Matrix is extremely inefficient for larger matrices as it explicitly computes all `N**2` determinants but we can do better using the characteristic polynomial:
```python
In [27]: A.charpoly()
Out[27]: PurePoly(1.0*lambda**2 + (-0.3*z - 1.5)*lambda - 0.03*z**2 + 0.13*z + 0.54, lambda, domain='RR[z]')

In [28]: a1, a2, deta = A.charpoly().all_coeffs()

In [29]: (-a1*A - a2*eye(2)) / deta
Out[29]: 
⎡       0.1⋅z + 0.6                -0.5⋅z - 0.8      ⎤
⎢─────────────────────────  ─────────────────────────⎥
⎢        2                          2                ⎥
⎢- 0.03⋅z  + 0.13⋅z + 0.54  - 0.03⋅z  + 0.13⋅z + 0.54⎥
⎢                                                    ⎥
⎢         -0.1⋅z                   0.2⋅z + 0.9       ⎥
⎢─────────────────────────  ─────────────────────────⎥
⎢        2                          2                ⎥
⎣- 0.03⋅z  + 0.13⋅z + 0.54  - 0.03⋅z  + 0.13⋅z + 0.54⎦
```
This method would only work for full-rank square systems but it does not rely on any polynomial cancellation. I don't know that it's possible to sensibly handle over-determined systems.

Maybe the solution for `RR(x)` and `CC(x)` is just to do like `solve(rational=True)`. The system could be converted to `QQ(x)` or `QQ_I(x)` internally for the purposes of `solve/rref` etc. Otherwise I think that polynomial calculations in the presence of rounding errors are probably not something that can be handled sensibly.

jksuom commented:
> polynomial calculations in the presence of rounding errors are probably not something that can be handled sensibly.

Agreed. I would also try converting `RR(x)` and `CC(x)` to `QQ(x)` and `QQ_I(x)`.
